File: math/149-max-points-on-a-line/MaxPointsOnALine.java
class Solution {
    public int maxPoints(int[][] points) {
        if(points.length<=2){
            return points.length;
        }
        int result = 0;
        for(int i=0;i<points.length;i++){
            Map<String, Integer> slopeMap = new HashMap<>();
            int duplicate = 1;
            for(int j=i+1;j<points.length;j++){
                int dx = points[j][0]-points[i][0];
                int dy = points[j][1]-points[i][1];

                if(dx == 0 && dy == 0){
                    duplicate++;
                    continue;
                }
                int g = gcd(dx,dy);
                dx/=g;
                dy/=g;
                String slope = dy+"/"+dx;
                slopeMap.put(slope, slopeMap.getOrDefault(slope, 0)+1);
            }
            int localMax = 0;
            for(int count:slopeMap.values()){
                localMax = Math.max(localMax, count);
            }
            result = Math.max(result, localMax + duplicate);
        }
        return result;
    }
    private int gcd(int a, int b){
        return b==0? a:gcd(b, a%b);
    }
}

File: math/149-max-points-on-a-line/MaxPointsOnALine.py
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        if len(points) <= 2:
            return len(points)

        max_points = 0

        for i in range(len(points)):
            slopes = defaultdict(int)
            duplicates = 1
            x1, y1 = points[i]

            for j in range(i + 1, len(points)):
                x2, y2 = points[j]

                if x1 == x2 and y1 == y2:
                    duplicates += 1
                else:
                    dx = x2 - x1
                    dy = y2 - y1
                    g = gcd(dx, dy)
                    dx //= g
                    dy //= g

                    # Always keep slope direction consistent
                    # We make sure the direction is unique by forcing dx to be positive
                    if dx < 0:
                        dx *= -1
                        dy *= -1

                    # Handle zero dx (vertical line)
                    if dx == 0:
                        dy = 1

                    # Handle zero dy (horizontal line)
                    if dy == 0:
                        dx = 1

                    slopes[(dy, dx)] += 1

            max_points = max(max_points, duplicates + max(slopes.values(), default=0))

        return max_points